### PR TITLE
Avoid allocations on lock in SyncCacheLock

### PR DESCRIPTION
**Pull request by bahusoid**

### CODE REVIEW

File: src/NHibernate/Cache/SyncCacheLock.cs [hunk 1]
```diff
@@ -6,34 +6,46 @@ namespace NHibernate.Cache
...
 		public IDisposable ReadLock()
 		{
-			return new MonitorLock(this);
+			return _monitorLock.Lock();
 		}
 
 		public IDisposable WriteLock()
 		{
-			return new MonitorLock(this);
+			return _monitorLock.Lock();
```
Comment (hazzik) [inline src/NHibernate/Cache/SyncCacheLock.cs:48]: But then it would be same lock for read and write. Is this expected?

Comment (bahusoid) [reply src/NHibernate/Cache/SyncCacheLock.cs:48]: Yes. It's the same lock for read and write without this change. This locker doesn't distinguish read and write.

Comment (hazzik) [reply src/NHibernate/Cache/SyncCacheLock.cs:48]: What about this?


	class SyncCacheLock : ICacheLock
	{
		public void Dispose()
		{
			Monitor.Exit(this);
		}

		public IDisposable ReadLock()
		{
			return Lock();
		}

		public IDisposable WriteLock()
		{
			return Lock();
		}

                private IDisposable Lock()
		{
                        Monitor.Enter(this);
			return this;
		}

                //....

**Review by hazzik — APPROVED** [review]

File: src/NHibernate/Cache/SyncCacheLock.cs (hunk 1 above)
Comment (bahusoid) [reply src/NHibernate/Cache/SyncCacheLock.cs:48]: No. ICacheLock is disposed separately from locks.

**Pull request merged**